File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.py
import collections
import heapq
import itertools


class Egg:
    def __init__(self, durability: int, weight: int):
        self.durability = durability
        self.weight = weight

    def crash_egg(self, other) -> None:
        self.durability, other.durability = (
            self.durability - other.weight, other.durability - self.weight
        )

    def crash_rollback(self, other) -> None:
        self.durability, other.durability = (
            self.durability + other.weight, other.durability + self.weight
        )

    def is_broken(self) -> bool:
        return self.durability <= 0


answer = 0


def dfs(left: int, broken_count: int):
    global eggs, answer, N
    if left == N:
        answer = max(answer, broken_count)
        return

    if eggs[left].is_broken() or broken_count == N - 1:
        dfs(left + 1, broken_count)
        return

    left_hand = eggs[left]
    for right in range(len(eggs)):
        if left == right:
            continue
        right_hand = eggs[right]
        if not right_hand.is_broken():
            left_hand.crash_egg(right_hand)
            broken = 0
            if left_hand.is_broken():
                broken += 1
            if right_hand.is_broken():
                broken += 1
            dfs(left + 1, broken_count + broken)
            left_hand.crash_rollback(eggs[right])


N = int(input())
eggs: list[Egg] = []

for _ in range(N):
    S, W = map(int, input().split())
    eggs.append(Egg(S, W))

dfs(0, 0)

print(answer)

File: 백준/Gold/16987. 계란으로 계란치기/계란으로 계란치기.java
import javax.naming.NameNotFoundException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static Egg[] eggs;
    static int answer;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        eggs = new Egg[N];

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            eggs[i] = new Egg(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()));
        }

        dfs(0, 0);

        System.out.println(answer);
    }

    static void dfs(int left, int broken) {
        if (left == eggs.length) {
            answer = Math.max(answer, broken);
            return;
        }

        if (eggs[left].isBroken() || broken == eggs.length - 1) {
            dfs(left + 1, broken);
            return;
        }

        Egg leftHand = eggs[left];

        for (int i = 0; i < eggs.length; i++) {
            if (i == left) continue;

            Egg rightHand = eggs[i];

            if (!rightHand.isBroken()) {
                leftHand.punch(rightHand);
                int brokenCount = 0;

                if (leftHand.isBroken()) {
                    brokenCount++;
                }
                if (rightHand.isBroken()) {
                    brokenCount++;
                }

                dfs(left + 1, broken + brokenCount);
                leftHand.rollbackPunch(rightHand);
            }
        }
    }

    static class Egg {
        private int S;
        private int W;

        public Egg(int s, int w) {
            S = s;
            W = w;
        }

        public void punch(Egg egg) {
            this.S -= egg.W;
            egg.S -= this.W;
        }

        public void rollbackPunch(Egg egg) {
            this.S += egg.W;
            egg.S += this.W;
        }

        public boolean isBroken() {
            return S <= 0;
        }
    }
}